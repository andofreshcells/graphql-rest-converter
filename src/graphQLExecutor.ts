import { execute, parse, GraphQLSchema, GraphQLError, ExecutionResult } from 'graphql'
import { ClientError, GraphQLClient } from 'graphql-request'

export type GraphQLExecutorArgs = {
  document: string
  variables: { [key: string]: any }
}

export type GraphQLExecutor = (args: GraphQLExecutorArgs) => Promise<ExecutionResult>

export const createHttpExecutor = (
  ...args: ConstructorParameters<typeof GraphQLClient>
): GraphQLExecutor => {
  const client = new GraphQLClient(...args)
  return async ({ document, variables }) => {
    try {
      // TODO: Use `errorPolicy: 'all'` when released
      return client.rawRequest({ query: document, variables })
    } catch (e) {
      if (e instanceof ClientError) {
        return e.response as ExecutionResult
      }
      throw e
    }
  }
}

export const createSchemaExecutor = (schema: GraphQLSchema): GraphQLExecutor => {
  return ({ document, variables }) =>
    Promise.resolve(execute({ schema, document: parse(document), variableValues: variables }))
}
