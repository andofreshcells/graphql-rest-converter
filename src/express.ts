/* eslint-disable @typescript-eslint/ban-ts-comment */
import _ from 'lodash'
import express, { RequestHandler, NextFunction, IRouter, Request, Response } from 'express'
import bodyParser from 'body-parser'
import { OpenAPIV3 } from 'openapi-types'
import { parse, buildSchema, print } from 'graphql'
import OpenAPIRequestCoercer from 'openapi-request-coercer'
import OpenAPIRequestValidator from 'openapi-request-validator'
import OpenAPIResponseValidator from 'openapi-response-validator'
import { getBridgeOperations } from './graphql.js'
import { pathTemplateToExpressRoute } from './pathTemplate.js'
import {
  BridgeOperation,
  BridgeOperations,
  SchemaComponents,
  CustomOperationProps,
  CreateOpenAPIGraphQLBridgeConfig,
  CreateOpenAPISchemaConfig,
  CreateMiddlewareConfig,
} from './types.js'
import { GraphQLExecutor } from './graphQLExecutor.js'
import RequestBodyObject = OpenAPIV3.RequestBodyObject
import { InvalidResponseError } from './errors.js'
import { createOpenAPISchemaWithValidate, resolveSchemaComponents } from './utils.js'
import { transformRequest } from './multipart.js'

const middlewareToPromise =
  (middleware: RequestHandler) =>
  (req: Parameters<RequestHandler>[0], res: Parameters<RequestHandler>[1]) => {
    return new Promise<ReturnType<RequestHandler>>((resolve, reject) => {
      const next = (x: unknown) => (x ? reject(x) : resolve())
      middleware(req, res, next)
    })
  }

const promiseToHandler =
  (
    handler: (
      req: Parameters<RequestHandler>[0],
      res: Parameters<RequestHandler>[1]
    ) => Promise<ReturnType<RequestHandler>>
  ) =>
  (...args: Parameters<RequestHandler>) => {
    const result = handler(args[0], args[1])
    result.catch(args[2] as NextFunction)
  }

const jsonBodyParserPromise = middlewareToPromise(bodyParser.json())
const formBodyParserPromise = middlewareToPromise(bodyParser.urlencoded())

const addOperation = <
  T extends CustomOperationProps = CustomOperationProps,
  Req extends Request = Request,
  Res extends Response = Response
>(
  router: IRouter,
  operation: BridgeOperation<T>,
  schemaComponents: SchemaComponents,
  executor: GraphQLExecutor<Req, Res>,
  config?: CreateMiddlewareConfig<Req, Res>
) => {
  const route = pathTemplateToExpressRoute(operation.path)

  // Resolving `$ref`, at least OpenAPIRequestValidator cannot handle them properly
  const parameters_ = _.cloneDeep(operation.openAPIOperation.parameters || [])
  const requestBody_ = _.cloneDeep(operation.openAPIOperation.requestBody) as
    | RequestBodyObject
    | undefined

  resolveSchemaComponents(parameters_, schemaComponents)
  resolveSchemaComponents(requestBody_, schemaComponents)

  const requestCoercer = new OpenAPIRequestCoercer.default({
    parameters: parameters_,
    requestBody: operation.requestBodyFormData === 'FORM_DATA' ? requestBody_ : undefined,
  })

  const requestValidator =
    typeof config?.validateRequest !== 'boolean' || config.validateRequest
      ? new OpenAPIRequestValidator.default({
          parameters: parameters_,
          requestBody: requestBody_,
        })
      : undefined

  const responseValidator = config?.validateResponse
    ? new OpenAPIResponseValidator.default({
        // Type in `openapi-response-validator` seems wrong
        responses: operation.openAPIOperation.responses as any,
        components: { schemas: schemaComponents },
      })
    : undefined

  const graphqlDocument_ = print(operation.graphqlDocument)

  router[operation.httpMethod](
    route,
    promiseToHandler(async (req, res) => {
      const allRequestBodyVariables = Object.keys(operation.requestBodyVariableMap)

      if (operation.requestBodyFormData === 'JSON') {
        await jsonBodyParserPromise(req, res)
      }

      if (operation.requestBodyFormData === 'FORM_DATA') {
        await formBodyParserPromise(req, res)
      }

      const supportedContentTypes = Object.keys(
        (operation.openAPIOperation.requestBody as OpenAPIV3.RequestBodyObject)?.content || {}
      )
      if (
        ['POST', 'PUT', 'PATCH'].includes(req.method) &&
        supportedContentTypes.length > 0 &&
        !supportedContentTypes.includes(req.headers['content-type'] || '')
      ) {
        res.status(415).json({
          errors: [
            {
              message: `Only "${supportedContentTypes.join(', ')}" supported`,
            },
          ],
        })
        return
      }
      const req_ = {
        ...req,
        cookies: { ...req.cookies },
        headers: { ...req.headers },
        params: { ...req.params },
        query: { ...req.query },
        body: req.body,
      }

      // TODO: Cookies not supported
      requestCoercer.coerce(req_)

      const validationResult = requestValidator ? requestValidator.validateRequest(req_) : null

      if (validationResult) {
        res.status(400).json({ errors: validationResult.errors })
        return
      }

      let variables: Record<string, unknown> = {}
      for (const parameter of operation.openAPIOperation
        .parameters as OpenAPIV3.ParameterObject[]) {
        const variableName = operation.variableMap[parameter.name] || parameter.name
        if (parameter.in === 'path') {
          variables[variableName] = req_.params[parameter.name]
        }
        if (parameter.in === 'query') {
          variables[variableName] = req_.query[parameter.name]
        }
        if (parameter.in === 'header') {
          variables[variableName] = req_.headers[parameter.name]
        }
      }
      if (operation.requestBodyFormData === 'JSON') {
        for (const requestBodyVariable of allRequestBodyVariables) {
          variables[requestBodyVariable] =
            req_.body[operation.requestBodyVariableMap[requestBodyVariable]] ||
            (operation.requestBodyIsSingleInput ? req_.body : null)
        }
      }

      if (operation.requestBodyFormData === 'FORM_DATA') {
        for (const requestBodyVariable of allRequestBodyVariables) {
          variables[requestBodyVariable] =
            req_.body[operation.requestBodyVariableMap[requestBodyVariable]] || null
        }
      }

      if (operation.requestBodyFormData === 'MULTIPART_FORM_DATA') {
        variables = {
          ...variables,
          ...transformRequest(
            req,
            graphqlDocument_,
            allRequestBodyVariables,
            operation.requestBodyVariableMap
          ),
        }
      }

      const request = {
        document: graphqlDocument_,
        variables,
        request: req as Req,
        response: res as Res,
      }
      const result = await executor(request)

      if (config?.responseTransformer) {
        const response_ = await config.responseTransformer({
          result,
          request,
          openAPISchema: {
            operation: operation.openAPIOperation,
            method: operation.httpMethod,
            path: operation.path,
          },
        })
        if (response_) {
          const { statusCode, contentType, data } = response_
          if (contentType) {
            res.contentType(contentType)
          }
          res.status(statusCode).send(data).end()
          return
        }
      }

      // There are 2 cases of errors handled here;

      // 1. If responseValidation is enabled, it will throw in both cases
      // - If there is a response validation problem
      // - If there are any graphql errors
      if (responseValidator) {
        const responseValidationErrors = responseValidator.validateResponse(200, result.data)
        if (responseValidationErrors || result.errors) {
          throw new InvalidResponseError(
            responseValidationErrors?.message || 'GraphQL Error',
            responseValidationErrors?.errors || [],
            result.errors
          )
        }
      }
      // 2. If no response validation is enabled, we will return (possibly) both the errors and a (partial) result
      if (result.errors) {
        res.status(500).json(result).end()
        return
      }
      res.status(200).json(result.data).end()
    })
  )
}

const createExpressMiddleware = <
  T extends CustomOperationProps = CustomOperationProps,
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response
>(
  operations: BridgeOperations<T>,
  executor: GraphQLExecutor<ThisRequest, ThisResponse>,
  config?: CreateMiddlewareConfig<ThisRequest, ThisResponse>
) => {
  // TODO: Avoid depending on express directly?
  const router = express.Router()

  for (const operation of operations.operations) {
    addOperation(router, operation, operations.schemaComponents, executor, config)
  }

  return router
}

export const createOpenAPIGraphQLBridge = <
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response,
  T extends CustomOperationProps = CustomOperationProps
>(
  config: CreateOpenAPIGraphQLBridgeConfig
) => {
  const { graphqlSchema, graphqlDocument, customScalars } = config

  const graphqlSchema_ =
    typeof graphqlSchema === 'string' ? buildSchema(graphqlSchema) : graphqlSchema

  const graphqlDocument_ =
    typeof graphqlDocument === 'string' ? parse(graphqlDocument) : graphqlDocument

  const operations = getBridgeOperations<T>(graphqlSchema_, graphqlDocument_, customScalars)

  return {
    getExpressMiddleware: (
      executor: GraphQLExecutor<ThisRequest, ThisResponse>,
      config?: CreateMiddlewareConfig
    ) => createExpressMiddleware(operations, executor, config),
    getOpenAPISchema: (config: CreateOpenAPISchemaConfig<T>): OpenAPIV3.Document<T> =>
      createOpenAPISchemaWithValidate<T>(operations, config),
  }
}
