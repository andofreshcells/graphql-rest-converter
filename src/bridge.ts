import _ from 'lodash'
import express, { RequestHandler, NextFunction, IRouter } from 'express'
import { OpenAPIV3 } from 'openapi-types'
import { parse } from 'graphql'
import { request as graphqlRequest } from 'graphql-request'
import { DocumentNode, GraphQLSchema } from 'graphql'
import OpenAPIRequestCoercer from 'openapi-request-coercer'
import OpenAPIRequestValidator, { OpenAPIRequestValidatorArgs } from 'openapi-request-validator'
import OpenAPIResponseValidator from 'openapi-response-validator'
import OpenAPISchemaValidator from 'openapi-schema-validator'
import { PartialDeep } from 'type-fest'
import { getOpenAPIGraphQLOperations } from './graphql'
import { pathTemplateToExpressRoute } from './pathTemplate'
import { createOpenAPISchemaFromOperations } from './openapi'
import { CustomProperties, OpenAPIGraphQLOperation } from './types'

export const asyncHandler =
  (handler: RequestHandler) =>
  (...args: Parameters<RequestHandler>) => {
    const result = handler(...args)
    return Promise.resolve(result).catch(args[args.length - 1] as NextFunction)
  }

const addOperation = (
  router: IRouter,
  operation: OpenAPIGraphQLOperation,
  config: CreateMiddlewareConfig
) => {
  const route = pathTemplateToExpressRoute(operation.path)

  const requestCoercer = new OpenAPIRequestCoercer({
    parameters: operation.openAPIOperation.parameters || [],
  })

  const requestValidator =
    typeof config.validateRequest !== 'boolean' || config.validateRequest
      ? new OpenAPIRequestValidator(operation.openAPIOperation as OpenAPIRequestValidatorArgs)
      : undefined

  const responseValidator = config.validateResponse
    ? // @ts-ignore
      new OpenAPIResponseValidator(operation.openAPIOperation)
    : undefined

  router[operation.httpMethod](
    route,
    asyncHandler(async (req, res) => {
      const reqClone = {
        ...req,
        headers: { ...req.headers },
        params: { ...req.params },
        query: { ...req.query },
      }
      requestCoercer.coerce(reqClone)
      const errors = requestValidator ? requestValidator.validateRequest(reqClone) : null
      if (errors) {
        res.status(400).json(errors)
        return
      }
      const parameters = { ...reqClone.params, ...reqClone.query }

      const data = await graphqlRequest({
        url: config.graphqlEndpoint,
        document: operation.graphqlDocument,
        variables: parameters,
        requestHeaders: {},
      })

      if (responseValidator) {
        const responseValidationErrors = responseValidator.validateResponse(200, data)
        if (responseValidationErrors) {
          throw new Error(JSON.stringify(responseValidationErrors))
        }
      }

      res.json(data)
    })
  )
}

export type CreateMiddlewareConfig = {
  graphqlEndpoint: string
  validateRequest?: boolean
  validateResponse?: boolean
}

export const createExpressMiddleware = (
  operations: Array<OpenAPIGraphQLOperation>,
  config: CreateMiddlewareConfig
) => {
  // TODO: Avoid depending on express directly?
  const router = express.Router()

  for (const operation of operations) {
    addOperation(router, operation, config)
  }

  return router
}

type OperationCustomProperties = {
  [CustomProperties.Operation]?: string
}

const getGraphQLOpenAPIOperationsFromOpenAPISchema = (
  schema: OpenAPIV3.Document<OperationCustomProperties>
) => {
  const operations: Array<OpenAPIGraphQLOperation> = []
  for (const [path, pathItem] of Object.entries(schema.paths)) {
    if (!pathItem) {
      continue
    }
    for (const [httpMethod, operation] of Object.entries(pathItem)) {
      // @ts-ignore
      if (!operation[CustomProperties.Operation]) {
        continue
      }
      operations.push({
        openAPIOperation: operation as OpenAPIV3.OperationObject,
        path,
        httpMethod: httpMethod as OpenAPIV3.HttpMethods,
        // @ts-ignore
        graphqlDocument: parse(operation[CustomProperties.Operation]),
      })
    }
  }
  return operations
}

export const createExpressMiddlewareFromOpenAPISchema = (
  schema: OpenAPIV3.Document<OperationCustomProperties>,
  config: CreateMiddlewareConfig
) => {
  const operations = getGraphQLOpenAPIOperationsFromOpenAPISchema(schema)
  return createExpressMiddleware(operations, config)
}

export const removeCustomProperties = (
  schema: OpenAPIV3.Document<OperationCustomProperties>
): OpenAPIV3.Document => {
  const schemaClone = _.cloneDeep(schema)
  for (const pathItem of Object.values(schemaClone.paths)) {
    if (!pathItem) {
      continue
    }
    for (const operation of Object.values(pathItem)) {
      if (typeof operation !== 'object') {
        continue
      }
      // @ts-ignore
      delete operation[CustomProperties.Operation]
    }
  }
  return schemaClone
}

export type CreateOpenAPISchemaConfig = {
  baseSchema: PartialDeep<OpenAPIV3.Document>
  validate?: boolean
}

const createOpenAPISchemaWithValidate = (
  operations: Array<OpenAPIGraphQLOperation>,
  config: CreateOpenAPISchemaConfig
) => {
  const openAPISchema = createOpenAPISchemaFromOperations(config.baseSchema, operations)
  if (config.validate) {
    const schemaValidator = new OpenAPISchemaValidator({ version: 3 })
    const schemaValidationErrors = schemaValidator.validate(openAPISchema)
    if (schemaValidationErrors?.errors?.length) {
      throw new Error(JSON.stringify(schemaValidationErrors))
    }
  }
  return openAPISchema
}

export type CreateOpenAPIGraphQLBridgeConfig = {
  graphqlSchema: GraphQLSchema
  graphqlDocument: DocumentNode
  customScalars?: (scalarTypeName: string) => OpenAPIV3.SchemaObject
}

export const createOpenAPIGraphQLBridge = (config: CreateOpenAPIGraphQLBridgeConfig) => {
  const { graphqlSchema, graphqlDocument, customScalars } = config

  const operations = getOpenAPIGraphQLOperations(graphqlSchema, graphqlDocument, customScalars)

  return {
    getExpressMiddleware: (config: CreateMiddlewareConfig) =>
      createExpressMiddleware(operations, config),
    getOpenAPISchema: (
      config: CreateOpenAPISchemaConfig
    ): OpenAPIV3.Document<OperationCustomProperties> =>
      createOpenAPISchemaWithValidate(operations, config),
  }
}
