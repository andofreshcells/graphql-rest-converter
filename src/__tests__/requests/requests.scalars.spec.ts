import express from 'express'
import { createSchemaExecutor } from '../../graphQLExecutor'
import request from 'supertest'
import { addMocksToSchema } from '@graphql-tools/mock'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { customScalarSchema, Scalars } from '../fixtures'
import { createOpenAPIGraphQLBridge } from '../../bridge'
import { gql } from 'graphql-tag'

const app = express()

export const gqlSchema = addMocksToSchema({
  mocks: {
    Date: () => '10-10-1985',
    Datetime: () => '10-10-1985:10:00:00',
    JSON: () => JSON.stringify({}),
  },
  schema: makeExecutableSchema({
    typeDefs: customScalarSchema,
  }),
})
const customScalarBridge = createOpenAPIGraphQLBridge({
  graphqlDocument: gql`
    query getData @OAOperation(path: "/data") {
      getData {
        id
        created
        data
        modified
      }
    }
    mutation storeData($input: DataInput!) @OAOperation(path: "/data") {
      storeData(data: $input) {
        id
        created
        modified
      }
    }
  `,
  graphqlSchema: customScalarSchema,
  customScalars: (name: string) => Scalars[name],
})

app.use(customScalarBridge.getExpressMiddleware(createSchemaExecutor(gqlSchema)))

describe('Requests with custom scalars', () => {
  it('should GET `/data`', async () => {
    const result = await request(app).get('/data')
    expect(result.body).toMatchSnapshot({
      getData: [
        {
          id: expect.any(Number),
        },
        {
          id: expect.any(Number),
        },
      ],
    })
  })
})
