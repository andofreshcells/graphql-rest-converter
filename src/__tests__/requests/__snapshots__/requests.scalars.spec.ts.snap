// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Requests with custom scalars should GET \`/data\` 1`] = `
{
  "getData": [
    {
      "created": "10-10-1985",
      "data": "{}",
      "id": Any<Number>,
      "modified": "10-10-1985:10:00:00",
    },
    {
      "created": "10-10-1985",
      "data": "{}",
      "id": Any<Number>,
      "modified": "10-10-1985:10:00:00",
    },
  ],
}
`;
